echoprometheus/README.md

```diff
@@ -0,0 +1,154 @@
+
+`NewMetric` function allows you to create custom metrics to be registed with default Prometheus registry. Never middleware does
+not provide any abstractions for metrics and guides you to use native Prometheus metrics and register them by yourself.
+
+This can be done newer middleware as follows:
+```go
+	e := echo.New()
+
+	customRegistry := prometheus.NewRegistry() // create custom registry for your custom metrics
+	customCounter := prometheus.NewCounter( // create new counter metric. This is replacement for `prometheus.Metric` struct
+		prometheus.CounterOpts{
+			Name: "custom_requests_total",
+			Help: "How many HTTP requests processed, partitioned by status code and HTTP method.",
+		},
+	)
+	if err := customRegistry.Register(customCounter); err != nil { // register your new counter metric with metrics registry
+		log.Fatal(err)
+	}
+
+	e.Use(NewMiddlewareWithConfig(MiddlewareConfig{
+		AfterNext: func(c echo.Context, err error) {
+			customCounter.Inc() // use our custom metric in middleware. after every request increment the counter
+		},
+		Registerer: customRegistry, // use our custom registry instead of default Prometheus registry
+	}))
+	e.GET("/metrics", NewHandlerWithConfig(HandlerConfig{Gatherer: customRegistry})) // register route for getting gathered metrics data from our custom Registry
+```
+
+## Replacement for `Prometheus.MetricsPath`
+
+`MetricsPath` was used to skip metrics own route from Prometheus metrics. Skipping is no longer done and requests to Prometheus
+route will be included in gathered metrics.
+
+This feature can be restored like that:
+```go
+conf := echoprometheus.MiddlewareConfig{
+    Skipper: func(c echo.Context) bool {
+        return c.Path() == "/metrics"
+    },
+}
+e.Use(echoprometheus.NewMiddlewareWithConfig(conf))
+```
+
+## Replacement for `Prometheus.RequestCounterURLLabelMappingFunc` and `Prometheus.RequestCounterHostLabelMappingFunc`
+
+These function fields were used to define how "URL" or "Host" attribute in Prometheus metric lines were created.
+
+These can be substituted by using `LabelFuncs`:
+```go
+	e.Use(echoprometheus.NewMiddlewareWithConfig(echoprometheus.MiddlewareConfig{
+		LabelFuncs: map[string]echoprometheus.LabelValueFunc{
+			"scheme": func(c echo.Context, err error) string { // additional custom label
+				return c.Scheme()
+			},
+			"url": func(c echo.Context, err error) string { // overrides default 'url' label value
+				return "x_" + c.Request().URL.Path
+			},
+			"host": func(c echo.Context, err error) string { // overrides default 'host' label value
+				return "y_" + c.Request().Host
+			},
+		},
+	}))
+```
+
+Will produce Prometheus line as
+`echo_request_duration_seconds_count{code="200",host="y_example.com",method="GET",scheme="http",url="x_/ok",scheme="http"} 1`
+
+
+## Replacement for `Metric.Buckets` and modifying default metrics
+
+Middleware register by default following metrics:
```

Comment:
Suggestion:
The `echoprometheus` middleware registers the following metrics by default: